src/hpr_routing.erl

```diff
@@ -97,95 +103,159 @@ find_routes({uplink, {_Type, DevAddr}}, PacketUp) ->
     ok = hpr_metrics:observe_find_routes(Time),
     Results.
 
+-spec start_crawl_routing() -> pid().
+start_crawl_routing() ->
+    erlang:spawn(fun() ->
+        Timer = timer:seconds(hpr_utils:get_env_int(routing_cleanup_timer_secs, 10)),
+        Window = timer:seconds(hpr_utils:get_env_int(routing_cleanup_window_secs, 60)),
+        ok = timer:sleep(Timer),
+        Removed = do_crawl_routing(Window),
+        lager:debug(
+            [{timer, Timer}, {window, Window}, {removed, Removed}],
+            "cleaning routing table"
+        ),
+        true = erlang:unregister(?ROUTING_CLEANUP),
+        true = erlang:register(?ROUTING_CLEANUP, start_crawl_routing())
+    end).
+
+-spec cancel_crawl_routing() -> ok.
+cancel_crawl_routing() ->
+    true = exit(get_crawl_routing_pid(), kill),
+    ok.
+
+-spec get_crawl_routing_pid() -> undefined | pid().
+get_crawl_routing_pid() ->
+    erlang:whereis(?ROUTING_CLEANUP).
+
+-spec do_crawl_routing(Window :: non_neg_integer()) -> NumDeleted :: non_neg_integer().
+do_crawl_routing(Window) ->
+    Now = erlang:system_time(millisecond) - Window,
+    %% MS = ets:fun2ms(fun({Hash, Time, Result}) when Time < Now -> true end),
+    MS = [{{'$1', '$2', '$3'}, [{'<', '$2', {const, Now}}], [true]}],
+    ets:select_delete(?PACKET_ETS, MS).
+
 %% ------------------------------------------------------------------
 %% Internal Function Definitions
 %% ------------------------------------------------------------------
 
+%% NOTE: all entries in `?PACKET_ETS' are 3-tuple {Hash, Time, Result} so it can
+%% be crawled with the same matchspec.
+-spec establish_routing(hpr_packet_up:packet(), Start :: non_neg_integer()) -> ok | {error, any()}.
+establish_routing(PacketUp, Start) ->
+    Hash = hpr_packet_up:phash(PacketUp),
+    true = ets:insert(?PACKET_ETS, {Hash, Start, {locked, [{PacketUp, Start}]}}),
```

Comment:
If you really want to only do this once, I would do an insert new here, look at `hpr_protocol_router:292`